src/audio/mfcc/mfcc_generic.c

```diff
@@ -0,0 +1,493 @@
+	int n1;
+	int n2;
+	int n;
+	int i;
+	int num_channels = source->channels;
+
+	/* Copy from source to pre-buffer for FFT.
+	 * The pre-emphasis filter is done in this step.
+	 */
+	for (copied = 0; copied < frames; copied += n) {
+		nmax = frames - copied;
+		n1 = audio_stream_frames_without_wrap(source, x);
+		n2 = mfcc_buffer_samples_without_wrap(buf, w);
+		n = MIN(n1, n2);
+		n = MIN(n, nmax);
+		x0 = x + source_channel;
+		for (i = 0; i < n; i++) {
+			if (emph->enable) {
+				/* Q1.15 x Q1.15 -> Q2.30 */
+				s = (int32_t)emph->delay * emph->coef + Q_SHIFT_LEFT(*x0, 15, 30);
+				*w = sat_int16(Q_SHIFT_RND(s, 30, 15));
+				emph->delay = *x0;
+			} else {
+				*w = *x0;
+			}
+			x0 += num_channels;
+			w++;
+		}
+
+		x = audio_stream_wrap(source, x + n * source->channels);
+		w = mfcc_buffer_wrap(buf, w);
+	}
+	buf->s_avail += copied;
+	buf->s_free -= copied;
+	buf->w_ptr = w;
+}
+
+static void mfcc_fill_prev_samples(struct mfcc_buffer *buf, int16_t *prev_data,
+				   int prev_data_length)
+{
+	/* Fill prev_data from input buffer */
+	int16_t *r = buf->r_ptr;
+	int16_t *p = prev_data;
+	int copied;
+	int nmax;
+	int n;
+
+	for (copied = 0; copied < prev_data_length; copied += n) {
+		nmax = prev_data_length - copied;
+		n = mfcc_buffer_samples_without_wrap(buf, r);
+		n = MIN(n, nmax);
+		memcpy(p, r, sizeof(int16_t) * n); /* Not using memcpy_s() due to speed need */
+		p += n;
+		r += n;
+		r = mfcc_buffer_wrap(buf, r);
+	}
+
+	buf->s_avail -= copied;
+	buf->s_free += copied;
+	buf->r_ptr = r;
+}
+
+static void mfcc_fill_fft_buffer(struct mfcc_state *state)
+{
+	struct mfcc_buffer *buf = &state->buf;
+	struct mfcc_fft *fft = &state->fft;
+	int16_t *r = buf->r_ptr;
+	int copied;
+	int nmax;
+	int idx = fft->fft_fill_start_idx;
+	int j;
+	int n;
+
+	/* Copy overlapped samples from state buffer. Imaginary part of input
+	 * remains zero.
+	 */
+	for (j = 0; j < state->prev_data_size; j++)
+		fft->fft_buf[idx + j].real = state->prev_data[j];
+
+	/* Copy hop size of new data from circular buffer */
+	idx += state->prev_data_size;
+	for (copied = 0; copied < fft->fft_hop_size; copied += n) {
+		nmax = fft->fft_hop_size - copied;
+		n = mfcc_buffer_samples_without_wrap(buf, r);
+		n = MIN(n, nmax);
+		for (j = 0; j < n; j++) {
+			fft->fft_buf[idx].real = *r;
+			r++;
+			idx++;
+		}
+		r = mfcc_buffer_wrap(buf, r);
+	}
+
+	buf->s_avail -= copied;
+	buf->s_free += copied;
+	buf->r_ptr = r;
+
+	/* Copy for next time data back to overlap buffer */
+	idx = fft->fft_fill_start_idx + fft->fft_hop_size;
+	for (j = 0; j < state->prev_data_size; j++)
+		state->prev_data[j] = fft->fft_buf[idx + j].real;
+}
+
+#ifdef MFCC_NORMALIZE_FFT
+static int mfcc_normalize_fft_buffer(struct mfcc_state *state)
+{
+	struct mfcc_fft *fft = &state->fft;
+	int32_t smax = 0;
+	int32_t x;
+	int shift;
+	int j;
+	int i = fft->fft_fill_start_idx;
+
+	for (j = 0; j < fft->fft_size; j++) {
+		x = fft->fft_buf[i + j].real;
+		smax = MAX(smax, x);
+		smax = MAX(smax, -x);
```

Comment:
is this a hot path? Then maybe reducing this to
```
	uint32_t absx = ABS(x);
	if (smax < absx)
		smax = absx;
```
would make sense

Reply:
Yes it's hot code so useful. I realized here's a possible overflow with -32768. Need another fix for norm_int32() ...

Reply:
Fix done, bit more complicated but should be safe now:

```
	for (j = 0; j < fft->fft_size; j++) {
		x = fft->fft_buf[i + j].real;
		absx = (x < 0) ? -x : x;
		if (smax < absx)
			smax = absx;
	}

	shift = norm_int32(smax << 15) - 1; /* 16 bit data */
	shift = MAX(shift, 0);
```